### ISSUE TITLE
KeepSprint bypasses when it could be avoided

### ISSUE THREAD

**Issue by ghost** [GrimAnticheat/Grim#1155]
### Describe the bypass and how to replicate it

When the player hits an entity, the simulation doesn't covers sprint being set to false, which causes keepsprint to bypass.

### Grim version

2.3.45

### Server version

git-PaperSpigot-445 (MC: 1.8.8)

### Plugins

Citzens, FastAssyncWorldEdit, PlayerCrasher, System, TChat, AdvancedBan, IPAuth, ViaVersion, LuckPerms, SkinsRestorer, WorldEdit, GrimAC, LPX.

**Comment by c0dingnoobi** [comment]
i dont think there is anyone rn willing or able to contribute the required changes withhin the prediction engine to flag keepsprint.
only thing which i could imagine is that someone would pull requests an additional check to mitigate keepsprint with buffers which is kind of unfit for something like grim

**Comment by ghost** [comment]
Well, hope some day it gets fixed. It would be better to make the changes in the prediction engine to avoid falses and be a more effective way to stop keepsprint, but overall it doesn't give a big advantage over the players, so it can wait.

**Comment by c0dingnoobi** [comment]
>  avoid falses

the problem with checking for keepsprint is that vanilla has a [sprint desync](https://github.com/GrimAnticheat/Grim/issues/838) which already flags grims prediction, if a keepsprint check will be implemented then that sprint desync will also affect the prediction of checking keepsprint. so yea it will false.

> hope some day it gets fixed

im not sure if 3.0 would check for attack slowdown or not, but that would the be best "hope" for now

> it doesn't give a big advantage

actually it does, but only when comboing an entity

**Comment by MWHunter** [comment]
sprinting is too fucked to check for this, this is not a prediction engine thing to check.  A check could use the prediction engine to supply statistics, though.

**Comment by c0dingnoobi** [comment]
i could pr a check based on buffers and logic using lastSprinting and isSprinting 
it might get messy but it would do its job

**Comment by SamB440** [comment]
> i could pr a check based on buffers and logic using lastSprinting and isSprinting it might get messy but it would do its job

If you hold the sprint key and attack an entity, the client doesn't send a sprinting change because you instantly start sprinting again. So I don't see how this could work. (Is 1.8 different?)

What exactly does keep sprint do beyond automatically sprinting when moving forward? Am I missing something? I see some modules remove the attack slowdown, which you could check for, but that would be a disadvantage as you take more kb?


Also my solution to the sprinting desync is just to split packet sprinting and attribute sprinting and loop the attribute. Nothing worked as well as that.

**Comment by c0dingnoobi** [comment]
> I see some modules remove the attack slowdown, which you could check for, but that would be a disadvantage as you take more kb?

exactly. keepsprint is mostly only an advantage when using it timed (while comboing a player), otherwhise it will just lead to gaining more knockback since you are removing the ability to reduce your knockback (due to having no attack slowdown)

**Comment by c0dingnoobi** [comment]
> Also my solution to the sprinting desync is just to split packet sprinting and attribute sprinting and loop the attribute. Nothing worked as well as that.

does this affect gameplay at all? 
i mean if not, and if it works reliably, would you want to pr ur changes?

**Comment by SamB440** [comment]
> > Also my solution to the sprinting desync is just to split packet sprinting and attribute sprinting and loop the attribute. Nothing worked as well as that.
> 
> does this affect gameplay at all? i mean if not, and if it works reliably, would you want to pr ur changes?

The biggest concern would be performance I guess. Since you're adding another possibility. Gameplay isn't affected.

I actually made this on a custom 1.18.2 simulation AC, so I will have to port the fix to Grim, and probably some attribute changes as those are handled a bit different, but I don't think it should be too hard.

**Comment by 0-x-2-2** [comment]
Isn't this actually possible to do in vanilla due to a bunch of bugs and the net code being fucked?

https://www.youtube.com/watch?v=NiOMVtrkV-M

**Comment by ghost** [comment]
> sprinting is too fucked to check for this, this is not a prediction engine thing to check. A check could use the prediction engine to supply statistics, though.

I don't think so... Keep sprint is not only allow you to keep sprint but avoid your motion to be slowed down. If you take a closer look at the vanilla code isn't that everytime you attack your motionXZ get *= 0.6D. Which is something that prediction engine can do.

**Comment by ghost** [comment]
I mean that if it just disable sprinting for a few ticks then players not going to care that much. It only annoying when it come to your motionXZ get slowed down by 0.6? Isn't that correct? Tho I'm not sure that happen when you on ground. If only seems to happened when you off ground but that still a huge advantage over other players. I mean it remove your abilities to reduce velocity too but that still have a huge advantage.

**Comment by ghost** [comment]
I think I could make add simple addition to the prediction engine to do this. Tho I don't know for sure that there is anyway to cause false expect taking another velocity.

**Comment by c0dingnoobi** [comment]
> everytime you attack your motionXZ get *= 0.6D. Which is something that prediction engine can do.

the issue with this is how grim tracks if a player sprints (since motion is only applied with 0.6D if the player sprints while attacking (attack slowdown)) 
thanks to mojang, that boolean desyncs 
check #379 for further explanations

> I think I could make add simple addition to the prediction engine to do this. Tho I don't know for sure that there is anyway to cause false expect taking another velocity.

ig no one would complain if you try

**Comment by 0-x-2-2** [comment]
This can be checked in the prediction engine on 1.8 but not on 1.9+ which is what Polar does as far as I know.

**Comment by ghost** [comment]
That's what define said; prediction engine could be used to supply the information needed to make a proper check for KeepSprint
